File: 0946-validate-stack-sequences/0946-validate-stack-sequences.cpp
// @Author: KING-SEN

class Solution {
public:
    bool validateStackSequences(vector<int>& pushed, vector<int>& popped) {
        stack<int> stack;
        int left = 0;
        for (int i = 0; i < pushed.size(); i++) {
            stack.push(pushed[i]);
            while (!stack.empty() && left < popped.size() && stack.top() == popped[left]) {
                left++;
                stack.pop();
            }
        }
        while (!stack.empty() && left < popped.size() && stack.top() == popped[left]) {
            left++;
            stack.pop();
        }
        return stack.empty() && left == popped.size();
    }
};

File: 0946-validate-stack-sequences/0946-validate-stack-sequences.py
class Solution:
    def validateStackSequences(self, pushed: List[int], popped: List[int]) -> bool:
        stack = []
        left = 0
        for right in range(len(pushed)):
            stack.append(pushed[right])
            while stack and stack[-1] == popped[left]:
                stack.pop()
                left += 1
        while stack and left < len(popped) and popped[left] == stack[-1]:
            left += 1
            stack.pop()
        return not stack and left == len(popped)